### ISSUE TITLE
Support full header context with read-only on encode phase

### ISSUE THREAD

**Issue by ardikabs** [envoyproxy/envoy#33873]
*Title*: *Support full header context with read-only on encode phase*

*Description*:
This request pertains specifically to the Golang extension, which by now if by any chance the Decode phase is skipped (unreached), thus when it comes to the Encode phase, there will be no information related to the request flowing during the Encode phase, such as but not limited to when we need to know which host associated to the response.

Requesting to add support for read-only header map during phases that have not yet been reached.

[optional *Relevant Links*:]
Slack thread: https://envoyproxy.slack.com/archives/C04QNSXC7U0/p1714279861985839

**Comment by doujiang24** [comment]
yep, we could add an `getRequestHeaders()` method in the `StreamInfo` interface.
just like the C++ API: https://github.com/envoyproxy/envoy/blob/main/envoy/stream_info/stream_info.h#L858

I'll have a try after https://github.com/envoyproxy/envoy/pull/33377

**Comment by doujiang24** [comment]
@phlax please help to add an `no stalebot` label, thanks~

**Comment by ardikabs** [comment]
now that #34810 has been merged, is this request still relevant @doujiang24?
